### PR TITLE
HPCC-13578 Prevent logging from destroyed crit secs from segfaulting

### PR DESCRIPTION
**Pull request by ghalliday**
Signed-off-by: Gavin Halliday gavin.halliday@lexisnexis.com

### CODE REVIEW

**Comment by ghalliday** [comment]
@richardkchapman please review.  Note, setting the manger to NULL may have an effect - worth double checking.

**Comment by AttilaVamos** [comment]
Automated Smoketest Build: success
HPCC Start: OK
HPCC Stop: OK

File: system/jlib/jlog.cpp [hunk 1]
```diff
@@ -2244,6 +2244,7 @@ MODULE_EXIT()
     delete thePassNoneFilter;
     delete thePassLocalFilter;
     delete thePassAllFilter;
+    theManager = NULL;
```
Comment (richardkchapman) [inline system/jlib/jlog.cpp]: What about the other ones?

**Comment by richardkchapman** [comment]
I can't see how setting it to NULL (rather than leaving it pointing to released memory) will make anything worse...

**Comment by ghalliday** [comment]
No nor I, but I thought I better point it out.

File: system/jlib/jmutex.hpp [hunk 2]
```diff
@@ -288,7 +288,15 @@ class CriticalSection
     inline ~CriticalSection()
     {
 #ifdef _ASSERT_LOCK_SUPPORT
-        assertex(owner==0 && depth==0);
```
Comment (richardkchapman) [inline system/jlib/jmutex.hpp]: This feels like it might be adding some overhead to a very commonly used bit of code...

**Comment by richardkchapman** [comment]
@ghalliday one minor comment

File: system/jlib/jmutex.hpp (hunk 2 above)
Comment (ghalliday) [inline system/jlib/jmutex.hpp]: On modern compilers try/catch should add no overhead to the code.
I could convert it to always use a fprintf instead, but potentially throwing an exception in a destructor isn't a good idea.

**Comment by ghalliday** [comment]
@richardkchapman which option do you prefer?

File: system/jlib/jmutex.hpp (hunk 2 above)
Comment (richardkchapman) [inline system/jlib/jmutex.hpp]: Perhaps it should just be an assert?

**Comment by richardkchapman** [comment]
@ghalliday back to you

File: system/jlib/jmutex.hpp (hunk 2 above)
Comment (ghalliday) [inline system/jlib/jmutex.hpp]: assert would only be triggered in debug mode (that may or may not be a good thing).
I have only ever hit this when interrupting a program.  I would probably go for a printf, but I'll implement any.  You can choose.

**Comment by richardkchapman** [comment]
I choose assert. That way we get the checking in debug builds without any overhead on release ones (since this is very commonly executed code)

**Comment by ghalliday** [comment]
@richardkchapman changed to assert, and cleared the other variables.

**Comment by AttilaVamos** [comment]
Automated Smoketest Build: success
HPCC Start: OK
HPCC Stop: OK

File: system/jlib/jexcept.cpp [hunk 3]
```diff
@@ -519,12 +518,17 @@ void raiseAssertException(const char *assertion, const char *file, unsigned line
...
+        printStackReport();
+        ERRLOG("%s",s.str());       // make sure doesn't get lost!
+        queryLogMsgManager()->flushQueue(10*1000);
 #ifdef _DEBUG
-    // cause a breakpoint in the debugger if we are debugging.
-    //userBreakpoint();
+        // cause a breakpoint in the debugger if we are debugging.
+        //userBreakpoint();
 #endif
+    }
```
Comment (richardkchapman) [inline system/jlib/jexcept.cpp]: Should there be an else fprintf(stderr...) ?

Comment (ghalliday) [inline system/jlib/jexcept.cpp]: I don't think so - nothing will have come out in the stack report either.

A good solution would be for DBGLOG to fallback to fprintf if the logMsgManager was NULL, but that would require a condition on every DBGLOG.  That might in itself be a good change.

**Pull request merged**